{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge American gut data\n",
    "> Scripts to process every OTU table I have and merge with metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install biom-format\n",
    "# !pip install anndata\n",
    "# !pip install scikit-bio\n",
    "# !pip install tqdm\n",
    "# !pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import biom\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import anndata\n",
    "# from skbio import TreeNode  # For tree\n",
    "from Bio import SeqIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_804250/1428360415.py:1: DtypeWarning: Columns (1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,23,24,25,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,44,47,48,49,50,51,52,53,54,55,56,57,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,119,120,121,122,123,124,125,126,127,129,130,131,132,133,134,135,136,138,139,140,141,142,143,144,145,146,147,148,150,151,152,153,154,155,156,157,158,159,168,169,170,171,173,174,175,176,177,178,179,180,181,182,183,184,185,186,188,189,190,191,192,193,194,195,196,197,199,201,202,203,204,205,206,207,208,209,210,211,212,213,214,216,217,218,219,220,221,222,223,225,226,227,228,229,230,231,232,233,234,235,236,237,238,240,241,242,243,244,245,246,247,248,249,251,252,253,254,256,257,258,259,260,261,262,263,264,267,268,269,270,272,273,274,275,276,277,279,282,284,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,320,321,322,323,324,325,326,327,328,329,331,332,334,335,336,337,338,339,340,341,342,344,345,346,347,348,349,350,351,352,353,354,356,357,358,359,360,361,362,363,365,366,367,368,369,370,371,372,373,374,375,376,380,381,382,383,384,385,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,408,409,410,411,412,413,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,465,466,467,468,469,470,471,473,474,475,477,478,479,480,481,482,483,484,485,486,487,489,490,491,492,493,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,524,525,526,527,528,529,530,531,532,533,534,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,658,659,660,662,663,665,666,667,668,669,670,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,690,693,694,695,696,697,698,699,700,701,702,703,705,706,707,708,709,710,712,713,714,715,716,717,718,719,720,721,722,724,725,726,727,728,729,730,731,732,733,734,735,736,738,739,740,741,742,743,744,745,746,747,748,749,750,753,754) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_table(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acid_reflux</th>\n",
       "      <th>acne_medication</th>\n",
       "      <th>acne_medication_otc</th>\n",
       "      <th>add_adhd</th>\n",
       "      <th>age_cat</th>\n",
       "      <th>age_years</th>\n",
       "      <th>alcohol_consumption</th>\n",
       "      <th>alcohol_frequency</th>\n",
       "      <th>alcohol_types_beercider</th>\n",
       "      <th>alcohol_types_red_wine</th>\n",
       "      <th>...</th>\n",
       "      <th>weekday_sleep_time</th>\n",
       "      <th>weekday_wake_time</th>\n",
       "      <th>weekend_sleep_time</th>\n",
       "      <th>weekend_wake_time</th>\n",
       "      <th>weight_cat</th>\n",
       "      <th>weight_change</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>weight_units</th>\n",
       "      <th>whole_eggs</th>\n",
       "      <th>whole_grain_frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10317.000001000</th>\n",
       "      <td>nan</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>diagnosed by a medical professional (doctor, p...</td>\n",
       "      <td>60s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>daily</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>remained stable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>never</td>\n",
       "      <td>occasionally (1-2 times/week)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10317.000001001</th>\n",
       "      <td>nan</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>nan</td>\n",
       "      <td>50s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>rarely (a few times/month)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>remained stable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10317.000001002</th>\n",
       "      <td>nan</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>regularly (3-5 times/week)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10317.000001004</th>\n",
       "      <td>nan</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>nan</td>\n",
       "      <td>40s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>rarely (a few times/month)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>remained stable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10317.000001008</th>\n",
       "      <td>nan</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>nan</td>\n",
       "      <td>60s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>never</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>increased more than 10 pounds</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 754 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                acid_reflux acne_medication acne_medication_otc  \\\n",
       "sample_name                                                       \n",
       "10317.000001000         nan              no                  no   \n",
       "10317.000001001         nan              no                  no   \n",
       "10317.000001002         nan              no                  no   \n",
       "10317.000001004         nan              no                  no   \n",
       "10317.000001008         nan              no                  no   \n",
       "\n",
       "                                                          add_adhd age_cat  \\\n",
       "sample_name                                                                  \n",
       "10317.000001000  diagnosed by a medical professional (doctor, p...     60s   \n",
       "10317.000001001                                                nan     50s   \n",
       "10317.000001002                                                nan     nan   \n",
       "10317.000001004                                                nan     40s   \n",
       "10317.000001008                                                nan     60s   \n",
       "\n",
       "                 age_years alcohol_consumption           alcohol_frequency  \\\n",
       "sample_name                                                                  \n",
       "10317.000001000        NaN                 yes                       daily   \n",
       "10317.000001001        NaN                 yes  rarely (a few times/month)   \n",
       "10317.000001002        NaN                 yes  regularly (3-5 times/week)   \n",
       "10317.000001004        NaN                 yes  rarely (a few times/month)   \n",
       "10317.000001008        NaN                  no                       never   \n",
       "\n",
       "                alcohol_types_beercider alcohol_types_red_wine  ...  \\\n",
       "sample_name                                                     ...   \n",
       "10317.000001000                    True                  False  ...   \n",
       "10317.000001001                   False                  False  ...   \n",
       "10317.000001002                   False                  False  ...   \n",
       "10317.000001004                   False                  False  ...   \n",
       "10317.000001008                   False                  False  ...   \n",
       "\n",
       "                weekday_sleep_time weekday_wake_time weekend_sleep_time  \\\n",
       "sample_name                                                               \n",
       "10317.000001000                nan               nan                nan   \n",
       "10317.000001001                nan               nan                nan   \n",
       "10317.000001002                nan               nan                nan   \n",
       "10317.000001004                nan               nan                nan   \n",
       "10317.000001008                nan               nan                nan   \n",
       "\n",
       "                weekend_wake_time weight_cat                  weight_change  \\\n",
       "sample_name                                                                   \n",
       "10317.000001000               nan        nan                remained stable   \n",
       "10317.000001001               nan        nan                remained stable   \n",
       "10317.000001002               nan        nan                            nan   \n",
       "10317.000001004               nan        nan                remained stable   \n",
       "10317.000001008               nan        nan  increased more than 10 pounds   \n",
       "\n",
       "                weight_kg weight_units whole_eggs  \\\n",
       "sample_name                                         \n",
       "10317.000001000       NaN          NaN      never   \n",
       "10317.000001001       NaN          NaN        nan   \n",
       "10317.000001002       NaN          NaN        nan   \n",
       "10317.000001004       NaN          NaN        nan   \n",
       "10317.000001008       NaN          NaN        nan   \n",
       "\n",
       "                         whole_grain_frequency  \n",
       "sample_name                                     \n",
       "10317.000001000  occasionally (1-2 times/week)  \n",
       "10317.000001001                            nan  \n",
       "10317.000001002                            nan  \n",
       "10317.000001004                            nan  \n",
       "10317.000001008                            nan  \n",
       "\n",
       "[5 rows x 754 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_table(\n",
    "    \"/home/phil/DATA/americangut/qiita_metadata.txt\", \n",
    "    dtype={0: str},\n",
    "    na_values=[\"not applicable\", \"missing\", \"unspecified\", \"unknown\", \"NaN\", \"nan\", \"N/A\", \"n/a\", \"LabControl test\", \"labcontrol test\"],\n",
    ")\n",
    "metadata = metadata.set_index(metadata.columns[0])\n",
    "\n",
    "# Clean up metadata: coerce mixed types to string\n",
    "for col in metadata.columns:\n",
    "    if metadata[col].dtype == np.dtype(\"O\"):\n",
    "        metadata[col] = metadata[col].astype(str)\n",
    "\n",
    "metadata.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'full_table.biom'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 18\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# For some reason, using biom.concat fails (non-disjoint index). This loop,\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[39m# while inelegant, works - go figure.\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[39m# with biom.util.biom_open(\"full_table.biom\", 'w') as f:\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[39m#     big_table.to_hdf5(f, \"merge_data.ipynb\")\u001b[39;00m\n\u001b[0;32m---> 18\u001b[0m big_table \u001b[39m=\u001b[39m biom\u001b[39m.\u001b[39;49mload_table(\u001b[39m\"\u001b[39;49m\u001b[39mfull_table.biom\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/biom/parse.py:666\u001b[0m, in \u001b[0;36mload_table\u001b[0;34m(f)\u001b[0m\n\u001b[1;32m    664\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m does not appear to be a BIOM file!\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m f)\n\u001b[1;32m    665\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 666\u001b[0m     \u001b[39mwith\u001b[39;49;00m biom_open(f) \u001b[39mas\u001b[39;49;00m fp:\n\u001b[1;32m    667\u001b[0m         \u001b[39mtry\u001b[39;49;00m:\n\u001b[1;32m    668\u001b[0m             table \u001b[39m=\u001b[39;49m parse_biom_table(fp)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/contextlib.py:137\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    135\u001b[0m \u001b[39mdel\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39margs, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mkwds, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfunc\n\u001b[1;32m    136\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 137\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mgen)\n\u001b[1;32m    138\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mStopIteration\u001b[39;00m:\n\u001b[1;32m    139\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mgenerator didn\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt yield\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mfrom\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/biom/util.py:434\u001b[0m, in \u001b[0;36mbiom_open\u001b[0;34m(fp, permission)\u001b[0m\n\u001b[1;32m    431\u001b[0m \u001b[39m# don't try to open an HDF5 file if H5PY is not installed, this can only\u001b[39;00m\n\u001b[1;32m    432\u001b[0m \u001b[39m# happen if we are reading a file\u001b[39;00m\n\u001b[1;32m    433\u001b[0m \u001b[39mif\u001b[39;00m mode \u001b[39min\u001b[39;00m {\u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mU\u001b[39m\u001b[39m'\u001b[39m}:\n\u001b[0;32m--> 434\u001b[0m     \u001b[39mif\u001b[39;00m os\u001b[39m.\u001b[39;49mpath\u001b[39m.\u001b[39;49mgetsize(fp) \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    435\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mThe file \u001b[39m\u001b[39m'\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m is empty and can\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt be parsed\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m fp)\n\u001b[1;32m    437\u001b[0m \u001b[39mif\u001b[39;00m mode \u001b[39min\u001b[39;00m [\u001b[39m'\u001b[39m\u001b[39mU\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m'\u001b[39m] \u001b[39mand\u001b[39;00m h5py\u001b[39m.\u001b[39mis_hdf5(fp):\n",
      "File \u001b[0;32m<frozen genericpath>:50\u001b[0m, in \u001b[0;36mgetsize\u001b[0;34m(filename)\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'full_table.biom'"
     ]
    }
   ],
   "source": [
    "# For some reason, using biom.concat fails (non-disjoint index). This loop,\n",
    "# while inelegant, works - go figure.\n",
    "\n",
    "# This is pretty slow, however:\n",
    "\n",
    "# dd = \"/home/phil/DATA/americangut/qiita_otus/\"\n",
    "# tables = [biom.load_table(f\"{dd}{x}\") for x in os.listdir(dd)]\n",
    "\n",
    "# big_table = tables[0]\n",
    "# it = tqdm(tables[1:], desc=f\"{big_table.shape}\")\n",
    "# for table in it:\n",
    "#     big_table = big_table.merge(table)\n",
    "#     it.set_description(f\"{big_table.shape}\")\n",
    "\n",
    "# with biom.util.biom_open(\"full_table.biom\", 'w') as f:\n",
    "#     big_table.to_hdf5(f, \"merge_data.ipynb\")\n",
    "\n",
    "big_table = biom.load_table(\"full_table.biom\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check that every OTU ID is in Greengenes:\n",
    "\n",
    "# gg_ids = set(\n",
    "#     [\n",
    "#         x.id\n",
    "#         for x in SeqIO.parse(\n",
    "#             \"/data/phil//greengenes/data/gg_13_5.fasta\", \"fasta\"\n",
    "#         )\n",
    "#     ]\n",
    "# )\n",
    "\n",
    "# otu_ids = set(big_table.ids(axis=\"observation\"))\n",
    "\n",
    "# print(len(otu_ids.difference(gg_ids)))\n",
    "# print(len(gg_ids.intersection(otu_ids)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "32608\n"
     ]
    }
   ],
   "source": [
    "# Now check that every sample ID is in the metadata:\n",
    "\n",
    "sample_ids = set(big_table.ids(axis=\"sample\"))\n",
    "\n",
    "missing = sample_ids.difference(set(metadata.index))\n",
    "print(len(missing))\n",
    "print(len(sample_ids.intersection(set(metadata.index))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create anndata object\n",
    "\n",
    "# Add metadata to table\n",
    "\n",
    "big_table.add_metadata(metadata.to_dict(orient=\"index\"), axis=\"sample\")\n",
    "adata = big_table.to_anndata()\n",
    "\n",
    "# # Add taxonomy:\n",
    "# tax = pd.read_table(\n",
    "#     \"/data/phil/greengenes/data/gg_13_5_taxonomy.txt\",\n",
    "#     dtype={0: str},\n",
    "#     header=None,\n",
    "# )\n",
    "# tax = tax.set_index(tax.columns[0])\n",
    "# tax = tax.loc[adata.var.index]\n",
    "# adata.var[\"tax_str\"] = tax\n",
    "# adata.var[\n",
    "#     [\"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"]\n",
    "# ] = tax[1].str.split(\";\", expand=True)\n",
    "\n",
    "# Normalize, save total counts:\n",
    "# adata.obs[\"total_counts\"] = adata.X.sum(axis=1)\n",
    "# adata.X /= adata.obs[\"total_counts\"].values.reshape(-1, 1)\n",
    "\n",
    "# # Add phylogeny:\n",
    "# tree = TreeNode.read(\n",
    "#     \"/home/phil/phylosig/greengenes/data/gg_13_5_otus/trees/97_otus.tree\"\n",
    "# )\n",
    "# tree = tree.shear(adata.var.index)\n",
    "# adata.var[\"phylogeny\"] = tree\n",
    "\n",
    "# # Add normalized data:\n",
    "# adata.layers[\"norm_counts\"] = adata.X / adata.X.sum(axis=1, keepdims=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"data/full_table.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[44], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# Save big_table:\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m big_table\u001b[39m.\u001b[39mto_tsv(\u001b[39m\"\u001b[39m\u001b[39mfull_table.tsv.gz\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mmerge_data.ipynb\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/IPython/core/displayhook.py:274\u001b[0m, in \u001b[0;36mDisplayHook.__call__\u001b[0;34m(self, result)\u001b[0m\n\u001b[1;32m    272\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwrite_format_data(format_dict, md_dict)\n\u001b[1;32m    273\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlog_output(format_dict)\n\u001b[0;32m--> 274\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfinish_displayhook()\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/ipykernel/displayhook.py:89\u001b[0m, in \u001b[0;36mZMQShellDisplayHook.finish_displayhook\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     87\u001b[0m sys\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mflush()\n\u001b[1;32m     88\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmsg[\u001b[39m\"\u001b[39m\u001b[39mcontent\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39m\"\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m\"\u001b[39m]:\n\u001b[0;32m---> 89\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msession\u001b[39m.\u001b[39msend(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpub_socket, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmsg, ident\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtopic)\n\u001b[1;32m     90\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmsg \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/jupyter_client/session.py:850\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, stream, msg_or_type, content, parent, ident, buffers, track, header, metadata)\u001b[0m\n\u001b[1;32m    848\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39madapt_version:\n\u001b[1;32m    849\u001b[0m     msg \u001b[39m=\u001b[39m adapt(msg, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39madapt_version)\n\u001b[0;32m--> 850\u001b[0m to_send \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mserialize(msg, ident)\n\u001b[1;32m    851\u001b[0m to_send\u001b[39m.\u001b[39mextend(buffers)\n\u001b[1;32m    852\u001b[0m longest \u001b[39m=\u001b[39m \u001b[39mmax\u001b[39m([\u001b[39mlen\u001b[39m(s) \u001b[39mfor\u001b[39;00m s \u001b[39min\u001b[39;00m to_send])\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/jupyter_client/session.py:719\u001b[0m, in \u001b[0;36mSession.serialize\u001b[0;34m(self, msg, ident)\u001b[0m\n\u001b[1;32m    717\u001b[0m     content \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnone\n\u001b[1;32m    718\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(content, \u001b[39mdict\u001b[39m):\n\u001b[0;32m--> 719\u001b[0m     content \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpack(content)\n\u001b[1;32m    720\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(content, \u001b[39mbytes\u001b[39m):\n\u001b[1;32m    721\u001b[0m     \u001b[39m# content is already packed, as in a relayed message\u001b[39;00m\n\u001b[1;32m    722\u001b[0m     \u001b[39mpass\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/site-packages/jupyter_client/session.py:95\u001b[0m, in \u001b[0;36mjson_packer\u001b[0;34m(obj)\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Convert a json object to a bytes.\"\"\"\u001b[39;00m\n\u001b[1;32m     94\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 95\u001b[0m     \u001b[39mreturn\u001b[39;00m json\u001b[39m.\u001b[39mdumps(\n\u001b[1;32m     96\u001b[0m         obj,\n\u001b[1;32m     97\u001b[0m         default\u001b[39m=\u001b[39mjson_default,\n\u001b[1;32m     98\u001b[0m         ensure_ascii\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m     99\u001b[0m         allow_nan\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    100\u001b[0m     )\u001b[39m.\u001b[39mencode(\u001b[39m\"\u001b[39m\u001b[39mutf8\u001b[39m\u001b[39m\"\u001b[39m, errors\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msurrogateescape\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    101\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mTypeError\u001b[39;00m, \u001b[39mValueError\u001b[39;00m) \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    102\u001b[0m     \u001b[39m# Fallback to trying to clean the json before serializing\u001b[39;00m\n\u001b[1;32m    103\u001b[0m     packed \u001b[39m=\u001b[39m json\u001b[39m.\u001b[39mdumps(\n\u001b[1;32m    104\u001b[0m         json_clean(obj),\n\u001b[1;32m    105\u001b[0m         default\u001b[39m=\u001b[39mjson_default,\n\u001b[1;32m    106\u001b[0m         ensure_ascii\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    107\u001b[0m         allow_nan\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    108\u001b[0m     )\u001b[39m.\u001b[39mencode(\u001b[39m\"\u001b[39m\u001b[39mutf8\u001b[39m\u001b[39m\"\u001b[39m, errors\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msurrogateescape\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/json/__init__.py:238\u001b[0m, in \u001b[0;36mdumps\u001b[0;34m(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)\u001b[0m\n\u001b[1;32m    232\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mcls\u001b[39m \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    233\u001b[0m     \u001b[39mcls\u001b[39m \u001b[39m=\u001b[39m JSONEncoder\n\u001b[1;32m    234\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mcls\u001b[39m(\n\u001b[1;32m    235\u001b[0m     skipkeys\u001b[39m=\u001b[39mskipkeys, ensure_ascii\u001b[39m=\u001b[39mensure_ascii,\n\u001b[1;32m    236\u001b[0m     check_circular\u001b[39m=\u001b[39mcheck_circular, allow_nan\u001b[39m=\u001b[39mallow_nan, indent\u001b[39m=\u001b[39mindent,\n\u001b[1;32m    237\u001b[0m     separators\u001b[39m=\u001b[39mseparators, default\u001b[39m=\u001b[39mdefault, sort_keys\u001b[39m=\u001b[39msort_keys,\n\u001b[0;32m--> 238\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkw)\u001b[39m.\u001b[39mencode(obj)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/json/encoder.py:200\u001b[0m, in \u001b[0;36mJSONEncoder.encode\u001b[0;34m(self, o)\u001b[0m\n\u001b[1;32m    196\u001b[0m         \u001b[39mreturn\u001b[39;00m encode_basestring(o)\n\u001b[1;32m    197\u001b[0m \u001b[39m# This doesn't pass the iterator directly to ''.join() because the\u001b[39;00m\n\u001b[1;32m    198\u001b[0m \u001b[39m# exceptions aren't as detailed.  The list call should be roughly\u001b[39;00m\n\u001b[1;32m    199\u001b[0m \u001b[39m# equivalent to the PySequence_Fast that ''.join() would do.\u001b[39;00m\n\u001b[0;32m--> 200\u001b[0m chunks \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39miterencode(o, _one_shot\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m    201\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(chunks, (\u001b[39mlist\u001b[39m, \u001b[39mtuple\u001b[39m)):\n\u001b[1;32m    202\u001b[0m     chunks \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39m(chunks)\n",
      "File \u001b[0;32m~/mambaforge/envs/americangut/lib/python3.11/json/encoder.py:258\u001b[0m, in \u001b[0;36mJSONEncoder.iterencode\u001b[0;34m(self, o, _one_shot)\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    254\u001b[0m     _iterencode \u001b[39m=\u001b[39m _make_iterencode(\n\u001b[1;32m    255\u001b[0m         markers, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdefault, _encoder, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mindent, floatstr,\n\u001b[1;32m    256\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mkey_separator, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mitem_separator, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msort_keys,\n\u001b[1;32m    257\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mskipkeys, _one_shot)\n\u001b[0;32m--> 258\u001b[0m \u001b[39mreturn\u001b[39;00m _iterencode(o, \u001b[39m0\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Save big_table:\n",
    "# big_table.to_tsv(\"full_table.tsv.gz\", \"merge_data.ipynb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = big_table.to_anndata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 32608 × 37215\n",
       "    obs: 'acid_reflux', 'acne_medication', 'acne_medication_otc', 'add_adhd', 'age_cat', 'age_years', 'alcohol_consumption', 'alcohol_frequency', 'alcohol_types_beercider', 'alcohol_types_red_wine', 'alcohol_types_sour_beers', 'alcohol_types_spiritshard_alcohol', 'alcohol_types_unspecified', 'alcohol_types_white_wine', 'allergic_to_i_have_no_food_allergies_that_i_know_of', 'allergic_to_peanuts', 'allergic_to_shellfish', 'allergic_to_tree_nuts', 'allergic_to_unspecified', 'alzheimers', 'animal_age', 'animal_free_text', 'animal_gender', 'animal_origin', 'animal_type', 'animal_type_free_text', 'anonymized_name', 'antibiotic_history', 'appendix_removed', 'artificial_gi_disorder_types_constipation', 'artificial_gi_disorder_types_diarrhea', 'artificial_gi_disorder_types_soft_stools', 'artificial_gi_disorder_types_stomachache', 'artificial_gi_disorder_types_unspecified', 'artificial_gi_disorders', 'artificial_sweeteners', 'artificial_sweeteners_food', 'asd', 'atypical_sleep_time', 'autoimmune', 'beet_frequency', 'birth_year', 'bmi', 'bmi_cat', 'body_product', 'body_site', 'bowel_movement_frequency', 'bowel_movement_quality', 'breastmilk_formula_ensure', 'cancer', 'cancer_treatment', 'cardiovascular_disease', 'cat', 'cdiff', 'census_region', 'chickenpox', 'clinical_condition', 'collection_timestamp', 'consume_animal_products_abx', 'contraceptive', 'coprophage', 'cosmetics_frequency', 'country', 'country_of_birth', 'country_residence', 'covid_anxious', 'covid_chronic_conditions_any_autoimmune_disease', 'covid_chronic_conditions_arthritis', 'covid_chronic_conditions_asthma_or_other_lung_problems', 'covid_chronic_conditions_atrial_fibrillation_or_atrial_flutter', 'covid_chronic_conditions_cancer', 'covid_chronic_conditions_chronic_obstructive_pulmonary_disease_copd', 'covid_chronic_conditions_congestive_heart_failure', 'covid_chronic_conditions_deep_vein_thrombosis', 'covid_chronic_conditions_diabetes_or_high_blood_sugar', 'covid_chronic_conditions_epilepsy_or_seizures', 'covid_chronic_conditions_frequent_or_very_bad_headaches', 'covid_chronic_conditions_heart_disease__myocardial_infarction', 'covid_chronic_conditions_heart_problems', 'covid_chronic_conditions_hiv', 'covid_chronic_conditions_hypertension', 'covid_chronic_conditions_immune_disorder', 'covid_chronic_conditions_kidney_problems', 'covid_chronic_conditions_pulmonary_embolism', 'covid_chronic_conditions_seasonal_allergies', 'covid_chronic_conditions_serious_acne_or_skin_problems', 'covid_chronic_conditions_serious_stomach_or_bowel_problems', 'covid_chronic_conditions_stroke', 'covid_chronic_conditions_unspecified', 'covid_depressed', 'covid_difficulty_falling_asleep', 'covid_difficulty_staying_asleep', 'covid_family_members_fallen_ill_physically', 'covid_family_members_hospitalized', 'covid_family_members_lost_job', 'covid_family_members_none_of_the_above', 'covid_family_members_passed_away', 'covid_family_members_put_into_self_quarantine_with_symptoms', 'covid_family_members_put_into_self_quarantine_without_symptoms_eg_due_to_possible_exposure', 'covid_family_members_reduced_ability_to_earn_money', 'covid_family_members_unspecified', 'covid_happened_to_you_fallen_ill_physically', 'covid_happened_to_you_hospitalized', 'covid_happened_to_you_lost_job', 'covid_happened_to_you_none_of_the_above', 'covid_happened_to_you_put_into_self_quarantine_with_symptoms', 'covid_happened_to_you_put_into_self_quarantine_without_symptoms_eg_due_to_possible_exposure', 'covid_happened_to_you_reduced_ability_to_earn_money', 'covid_happened_to_you_unspecified', 'covid_interest_pleasure', 'covid_left_home', 'covid_level_of_wellbeing', 'covid_likely_exposure_no', 'covid_likely_exposure_unspecified', 'covid_likely_exposure_yes_someone_with_medical_diagnosis_but_no_test', 'covid_likely_exposure_yes_someone_with_positive_test', 'covid_likely_exposure_yes_someone_with_possible_symptoms_but_no_diagnosis_by_doctor', 'covid_occupation', 'covid_patient_care', 'covid_patient_care_ppe', 'covid_patient_care_seven_days', 'covid_patient_care_seven_days_confirmed', 'covid_quality_of_life', 'covid_rideshare', 'covid_sleep_interference', 'covid_sleep_pattern', 'covid_suspected_positive', 'covid_suspected_positive_date', 'covid_symptoms_chest_pain_or_tightness_in_chest', 'covid_symptoms_cough', 'covid_symptoms_diarrhea', 'covid_symptoms_fatigue', 'covid_symptoms_fever', 'covid_symptoms_lack_of_appetitie', 'covid_symptoms_loss_of_taste_or_smell', 'covid_symptoms_nausea', 'covid_symptoms_other', 'covid_symptoms_shortness_of_breath', 'covid_symptoms_sore_throat', 'covid_symptoms_stay_home', 'covid_symptoms_unspecified', 'covid_waking_up_early', 'covid_worried_sleep', 'covid_worrying', 'csection', 'dark_mode_on', 'deodorant_use', 'depression_bipolar_schizophrenia', 'description', 'diabetes', 'diabetes_type', 'diet', 'diet_type', 'dna_extracted', 'dog', 'dominant_hand', 'drinking_water_source', 'drinks_per_session', 'economic_region', 'elevation', 'empo_1', 'empo_2', 'empo_3', 'env_biome', 'env_feature', 'env_material', 'env_package', 'epilepsy_or_seizure_disorder', 'exercise_frequency', 'exercise_location', 'fed_as_infant', 'fermented_consumed', 'fermented_consumed_beer', 'fermented_consumed_chicha', 'fermented_consumed_cider', 'fermented_consumed_cottage_cheese', 'fermented_consumed_fermented_beansmisonatto', 'fermented_consumed_fermented_breadsourdoughinjera', 'fermented_consumed_fermented_fish', 'fermented_consumed_fermented_tofu', 'fermented_consumed_fish_sauce', 'fermented_consumed_kefir_milk', 'fermented_consumed_kefir_water', 'fermented_consumed_kimchi', 'fermented_consumed_kombucha', 'fermented_consumed_mead', 'fermented_consumed_other', 'fermented_consumed_other_1', 'fermented_consumed_pickled_vegetables', 'fermented_consumed_sauerkraut', 'fermented_consumed_sour_creamcreme_fraiche', 'fermented_consumed_tempeh', 'fermented_consumed_unspecified', 'fermented_consumed_wine', 'fermented_consumed_yogurtlassi', 'fermented_frequency', 'fermented_increased', 'fermented_other', 'fermented_plant_frequency', 'fermented_produce_commercial', 'fermented_produce_commercial_beer', 'fermented_produce_commercial_chicha', 'fermented_produce_commercial_cider', 'fermented_produce_commercial_cottage_cheese', 'fermented_produce_commercial_fermented_beansmisonatto', 'fermented_produce_commercial_fermented_breadsourdoughinjera', 'fermented_produce_commercial_fermented_fish', 'fermented_produce_commercial_fermented_tofu', 'fermented_produce_commercial_fish_sauce', 'fermented_produce_commercial_kefir_milk', 'fermented_produce_commercial_kefir_water', 'fermented_produce_commercial_kimchi', 'fermented_produce_commercial_kombucha', 'fermented_produce_commercial_mead', 'fermented_produce_commercial_other', 'fermented_produce_commercial_other_1', 'fermented_produce_commercial_pickled_vegetables', 'fermented_produce_commercial_sauerkraut', 'fermented_produce_commercial_sour_creamcreme_fraiche', 'fermented_produce_commercial_tempeh', 'fermented_produce_commercial_unspecified', 'fermented_produce_commercial_wine', 'fermented_produce_commercial_yogurtlassi', 'fermented_produce_personal', 'fermented_produce_personal_beer', 'fermented_produce_personal_chicha', 'fermented_produce_personal_cider', 'fermented_produce_personal_cottage_cheese', 'fermented_produce_personal_fermented_beansmisonatto', 'fermented_produce_personal_fermented_breadsourdoughinjera', 'fermented_produce_personal_fermented_fish', 'fermented_produce_personal_fermented_tofu', 'fermented_produce_personal_fish_sauce', 'fermented_produce_personal_kefir_milk', 'fermented_produce_personal_kefir_water', 'fermented_produce_personal_kimchi', 'fermented_produce_personal_kombucha', 'fermented_produce_personal_mead', 'fermented_produce_personal_other', 'fermented_produce_personal_other_1', 'fermented_produce_personal_pickled_vegetables', 'fermented_produce_personal_sauerkraut', 'fermented_produce_personal_sour_creamcreme_fraiche', 'fermented_produce_personal_tempeh', 'fermented_produce_personal_unspecified', 'fermented_produce_personal_wine', 'fermented_produce_personal_yogurtlassi', 'flossing_frequency', 'flu_vaccine_date', 'food_source', 'food_source_human_food', 'food_source_pet_store_food', 'food_source_unspecified', 'food_source_wild_food', 'food_special', 'food_special_grain_free', 'food_special_organic', 'food_special_unspecified', 'food_type', 'frozen_dessert_frequency', 'fruit_frequency', 'fungal_overgrowth', 'geo_loc_name', 'gluten', 'height_cm', 'height_units', 'high_fat_red_meat_frequency', 'homecooked_meals_frequency', 'host_age', 'host_age_normalized_years', 'host_age_units', 'host_body_habitat', 'host_body_mass_index', 'host_body_product', 'host_body_site', 'host_common_name', 'host_height', 'host_height_units', 'host_scientific_name', 'host_subject_id', 'host_taxid', 'host_weight', 'host_weight_units', 'hours_outside', 'humans_free_text', 'ibd', 'ibd_diagnosis', 'ibd_diagnosis_refined', 'ibs', 'keep_sample_for_thdmi', 'kidney_disease', 'lactose', 'last_move', 'last_travel', 'latitude', 'level_of_education', 'lifestage', 'liver_disease', 'living_status', 'livingwith', 'longitude', 'lowgrain_diet_type', 'lung_disease', 'meat_eggs_frequency', 'mental_illness', 'mental_illness_type_anorexia_nervosa', 'mental_illness_type_bipolar_disorder', 'mental_illness_type_bulimia_nervosa', 'mental_illness_type_depression', 'mental_illness_type_ptsd_post_traumatic_stress_disorder', 'mental_illness_type_schizophrenia', 'mental_illness_type_substance_abuse', 'mental_illness_type_unspecified', 'migraine', 'milk_cheese_frequency', 'milk_substitute_frequency', 'multivitamin', 'nail_biter', 'name', 'non_food_allergies_beestings', 'non_food_allergies_drug_eg_penicillin', 'non_food_allergies_pet_dander', 'non_food_allergies_poison_ivyoak', 'non_food_allergies_sun', 'non_food_allergies_unspecified', 'oils_frequency_oxalate', 'oils_frequency_soy', 'olive_oil', 'one_liter_of_water_a_day_frequency', 'other_animals_free_text', 'other_supplement_frequency', 'pets_other', 'physical_specimen_location', 'physical_specimen_remaining', 'pku', 'plant_protein_frequency', 'pm_age', 'pm_concern', 'pm_education', 'pm_email', 'pm_ethnicity', 'pm_experience_worth_it', 'pm_gastro_problems', 'pm_gastro_problems_crohns_disease_or_ulcerative_colitis', 'pm_gastro_problems_frequent_more_than_once_a_week_constipation', 'pm_gastro_problems_frequent_more_than_once_a_week_diarrhea', 'pm_gastro_problems_gastrointestinal_cancer', 'pm_gastro_problems_i_have_had_surgery_on_my_intestines', 'pm_gastro_problems_irritable_bowel_syndrome_ibs', 'pm_gastro_problems_other', 'pm_gastro_problems_unspecified', 'pm_gender', 'pm_health', 'pm_lifestyle_change', 'pm_lifestyle_change_how', 'pm_lifestyle_change_how_change_in_alcohol_or_tobacco_use_ie_starting_stopping_or_changing_amount_consumed', 'pm_lifestyle_change_how_change_in_diet_ie_taking_a_probiotic', 'pm_lifestyle_change_how_change_in_exercise', 'pm_lifestyle_change_how_change_in_physical_environment_ie_adopting_or_getting_rid_of_a_pet_cleaning_more_or_less_often', 'pm_lifestyle_change_how_change_in_use_of_nutritional_supplements_or_vitamins', 'pm_lifestyle_change_how_other', 'pm_lifestyle_change_how_unspecified', 'pm_name', 'pm_participation_reason', 'pm_participation_reason_desire_to_improve_my_health', 'pm_participation_reason_general_curiosityinterest_in_learning_about_my_microbiome', 'pm_participation_reason_i_have_gastrointestinal_problems', 'pm_participation_reason_i_have_other_microbiomerelevant_health_problems', 'pm_participation_reason_interest_in_contributing_to_science', 'pm_participation_reason_other', 'pm_participation_reason_people_in_my_family_or_social_network_have_done_it', 'pm_participation_reason_professional_interest_in_the_microbiome', 'pm_participation_reason_the_study_seemed_fun_and_entertaining', 'pm_participation_reason_unspecified', 'pm_recontact', 'pm_shared', 'pm_shared_pcp', 'pm_shared_pcp_changes', 'pm_shared_pcp_changes_what', 'pm_shared_pcp_changes_what_change_in_behaviorallifestyle_recommendations_eg_start_probiotic', 'pm_shared_pcp_changes_what_change_in_medication_ie_starting_stopping_or_changing_the_dose_of_a_prescription', 'pm_shared_pcp_changes_what_ordered_additional_labs_tests_or_procedures', 'pm_shared_pcp_changes_what_other', 'pm_shared_pcp_changes_what_referral_to_a_specialist_eg_gastroenterologist', 'pm_shared_pcp_changes_what_unspecified', 'pm_shared_who', 'pm_shared_who_colleagues', 'pm_shared_who_family_members', 'pm_shared_who_friends', 'pm_shared_who_nutritionistdietician', 'pm_shared_who_other', 'pm_shared_who_other_medical_or_health_professional', 'pm_shared_who_posteddiscussed_on_an_online_patienthealth_platform_ie_patientslikeme', 'pm_shared_who_posteddiscussed_on_data_sharing_platform_ie_open_humans', 'pm_shared_who_posteddiscussed_on_social_networking_site_ie_facebook', 'pm_shared_who_primary_care_physician', 'pm_shared_who_specialty_physician_eg_gastroenterologist', 'pm_shared_who_unspecified', 'pm_understand_results', 'pm_useful', 'pool_frequency', 'poultry_frequency', 'pregnant', 'prepared_meals_frequency', 'probiotic_frequency', 'public', 'qiita_study_id', 'race', 'ready_to_eat_meals_frequency', 'red_meat_frequency', 'roommates', 'roommates_in_study', 'salted_snacks_frequency', 'sample_type', 'scientific_name', 'seafood_frequency', 'seasonal_allergies', 'setting', 'sex', 'sibo', 'skin_condition', 'sleep_duration', 'sleep_quality', 'smoking_frequency', 'softener', 'specialized_diet_exclude_dairy', 'specialized_diet_exclude_nightshades', 'specialized_diet_exclude_refined_sugars', 'specialized_diet_fodmap', 'specialized_diet_halaal', 'specialized_diet_i_do_not_eat_a_specialized_diet', 'specialized_diet_kosher', 'specialized_diet_modified_paleo_diet', 'specialized_diet_other_restrictions_not_described_here', 'specialized_diet_paleo_diet_or_primal_diet', 'specialized_diet_raw_food_diet', 'specialized_diet_unspecified', 'specialized_diet_weston_price_or_other_low_grain_low_processed_food_diet', 'state', 'subset_age', 'subset_antibiotic_history', 'subset_bmi', 'subset_diabetes', 'subset_healthy', 'subset_ibd', 'sugar_sweetened_drink_frequency', 'sugary_sweets_frequency', 'surf_board_type', 'surf_frequency', 'surf_loal_break_frequency', 'surf_local_break', 'surf_shower_frequency', 'surf_stance', 'surf_sunscreen', 'surf_sunscreen_frequency', 'surf_travel_distance', 'surf_travel_frequency', 'surf_wax', 'surf_weetsuit', 'taxon_id', 'teethbrushing_frequency', 'thdmi_cohort', 'thyroid', 'title', 'toilet_water_access', 'tonsils_removed', 'types_of_plants', 'vegetable_frequency', 'vioscreen_a_bev', 'vioscreen_a_cal', 'vioscreen_acesupot', 'vioscreen_activity_level', 'vioscreen_add_sug', 'vioscreen_addsugar', 'vioscreen_adsugtot', 'vioscreen_age', 'vioscreen_alanine', 'vioscreen_alcohol', 'vioscreen_alcohol_servings', 'vioscreen_alphacar', 'vioscreen_alphtoce', 'vioscreen_alphtoco', 'vioscreen_arginine', 'vioscreen_ash', 'vioscreen_aspartam', 'vioscreen_aspartic', 'vioscreen_avcarb', 'vioscreen_bcodeid', 'vioscreen_betacar', 'vioscreen_betacryp', 'vioscreen_betaine', 'vioscreen_betatoco', 'vioscreen_biochana', 'vioscreen_bmi', 'vioscreen_caffeine', 'vioscreen_calcium', 'vioscreen_calcium_avg', 'vioscreen_calcium_dose', 'vioscreen_calcium_freq', 'vioscreen_calcium_from_dairy_servings', 'vioscreen_calcium_servings', 'vioscreen_calories', 'vioscreen_carbo', 'vioscreen_cholest', 'vioscreen_choline', 'vioscreen_clac9t11', 'vioscreen_clat10c12', 'vioscreen_copper', 'vioscreen_coumest', 'vioscreen_cystine', 'vioscreen_d_cheese', 'vioscreen_d_milk', 'vioscreen_d_tot_soym', 'vioscreen_d_total', 'vioscreen_d_yogurt', 'vioscreen_daidzein', 'vioscreen_database', 'vioscreen_delttoco', 'vioscreen_discfat_oil', 'vioscreen_discfat_sol', 'vioscreen_dob', 'vioscreen_eer', 'vioscreen_email', 'vioscreen_erythr', 'vioscreen_f_citmlb', 'vioscreen_f_nj_citmlb', 'vioscreen_f_nj_other', 'vioscreen_f_nj_total', 'vioscreen_f_other', 'vioscreen_f_total', 'vioscreen_fat', 'vioscreen_fiber', 'vioscreen_fibh2o', 'vioscreen_fibinso', 'vioscreen_finished', 'vioscreen_fish_servings', 'vioscreen_fol_deqv', 'vioscreen_fol_nat', 'vioscreen_fol_syn', 'vioscreen_formontn', 'vioscreen_fried_fish_servings', 'vioscreen_fried_food_servings', 'vioscreen_frt5_day', 'vioscreen_frtsumm', 'vioscreen_fructose', 'vioscreen_fruit_servings', 'vioscreen_g_nwhl', 'vioscreen_g_total', 'vioscreen_g_whl', 'vioscreen_galactos', 'vioscreen_gammtoco', 'vioscreen_gender', 'vioscreen_genistn', 'vioscreen_glac', 'vioscreen_gltc', 'vioscreen_glucose', 'vioscreen_glutamic', 'vioscreen_glycine', 'vioscreen_glycitn', 'vioscreen_grams', 'vioscreen_hei2010_dairy', 'vioscreen_hei2010_empty_calories', 'vioscreen_hei2010_fatty_acids', 'vioscreen_hei2010_fruit', 'vioscreen_hei2010_greens_beans', 'vioscreen_hei2010_protien_foods', 'vioscreen_hei2010_refined_grains', 'vioscreen_hei2010_score', 'vioscreen_hei2010_sea_foods_plant_protiens', 'vioscreen_hei2010_sodium', 'vioscreen_hei2010_veg', 'vioscreen_hei2010_whole_fruit', 'vioscreen_hei2010_whole_grains', 'vioscreen_hei_drk_g_org_veg_leg', 'vioscreen_hei_fruit', 'vioscreen_hei_grains', 'vioscreen_hei_meat_beans', 'vioscreen_hei_milk', 'vioscreen_hei_non_juice_frt', 'vioscreen_hei_oils', 'vioscreen_hei_sat_fat', 'vioscreen_hei_score', 'vioscreen_hei_sodium', 'vioscreen_hei_sol_fat_alc_add_sug', 'vioscreen_hei_veg', 'vioscreen_hei_whl_grains', 'vioscreen_height', 'vioscreen_histidin', 'vioscreen_inositol', 'vioscreen_iron', 'vioscreen_isoleuc', 'vioscreen_isomalt', 'vioscreen_joules', 'vioscreen_juice_servings', 'vioscreen_lactitol', 'vioscreen_lactose', 'vioscreen_legumes', 'vioscreen_leucine', 'vioscreen_line_gi', 'vioscreen_low_fat_dairy_serving', 'vioscreen_lutzeax', 'vioscreen_lycopene', 'vioscreen_lysine', 'vioscreen_m_egg', 'vioscreen_m_fish_hi', 'vioscreen_m_fish_lo', 'vioscreen_m_frank', 'vioscreen_m_meat', 'vioscreen_m_mpf', 'vioscreen_m_nutsd', 'vioscreen_m_organ', 'vioscreen_m_poult', 'vioscreen_m_soy', 'vioscreen_magnes', 'vioscreen_maltitol', 'vioscreen_maltose', 'vioscreen_mangan', 'vioscreen_mannitol', 'vioscreen_methhis3', 'vioscreen_methion', 'vioscreen_mfa141', 'vioscreen_mfa161', 'vioscreen_mfa181', 'vioscreen_mfa201', 'vioscreen_mfa221', 'vioscreen_mfatot', 'vioscreen_multi_calcium_avg', 'vioscreen_multi_calcium_dose', 'vioscreen_multivitamin', 'vioscreen_multivitamin_freq', 'vioscreen_natoco', 'vioscreen_nccglbr', 'vioscreen_nccglgr', 'vioscreen_niacin', 'vioscreen_niacineq', 'vioscreen_nitrogen', 'vioscreen_non_fried_fish_servings', 'vioscreen_nutrient_recommendation', 'vioscreen_omega3', 'vioscreen_oxalic', 'vioscreen_oxalicm', 'vioscreen_pantothe', 'vioscreen_pectins', 'vioscreen_pfa182', 'vioscreen_pfa183', 'vioscreen_pfa184', 'vioscreen_pfa204', 'vioscreen_pfa205', 'vioscreen_pfa225', 'vioscreen_pfa226', 'vioscreen_pfatot', 'vioscreen_phenylal', 'vioscreen_phosphor', 'vioscreen_phytic', 'vioscreen_pinitol', 'vioscreen_potass', 'vioscreen_procdate', 'vioscreen_proline', 'vioscreen_protanim', 'vioscreen_protein', 'vioscreen_protocol', 'vioscreen_protveg', 'vioscreen_questionnaire', 'vioscreen_recno', 'vioscreen_retinol', 'vioscreen_rgrain', 'vioscreen_ribofla', 'vioscreen_sacchar', 'vioscreen_salad_vegetable_servings', 'vioscreen_satoco', 'vioscreen_scf', 'vioscreen_scfv', 'vioscreen_selenium', 'vioscreen_serine', 'vioscreen_sfa100', 'vioscreen_sfa120', 'vioscreen_sfa140', 'vioscreen_sfa160', 'vioscreen_sfa170', 'vioscreen_sfa180', 'vioscreen_sfa200', 'vioscreen_sfa220', 'vioscreen_sfa40', 'vioscreen_sfa60', 'vioscreen_sfa80', 'vioscreen_sfatot', 'vioscreen_sodium', 'vioscreen_sorbitol', 'vioscreen_srvid', 'vioscreen_starch', 'vioscreen_started', 'vioscreen_subject_id', 'vioscreen_sucpoly', 'vioscreen_sucrlose', 'vioscreen_sucrose', 'vioscreen_sweet_servings', 'vioscreen_tagatose', 'vioscreen_tfa161t', 'vioscreen_tfa181t', 'vioscreen_tfa182t', 'vioscreen_tgrain', 'vioscreen_thiamin', 'vioscreen_threonin', 'vioscreen_time', 'vioscreen_totaltfa', 'vioscreen_totcla', 'vioscreen_totfolat', 'vioscreen_totsugar', 'vioscreen_tryptoph', 'vioscreen_tyrosine', 'vioscreen_user_id', 'vioscreen_v_drkgr', 'vioscreen_v_orange', 'vioscreen_v_other', 'vioscreen_v_potato', 'vioscreen_v_starcy', 'vioscreen_v_tomato', 'vioscreen_v_total', 'vioscreen_valine', 'vioscreen_veg5_day', 'vioscreen_vegetable_servings', 'vioscreen_vegsumm', 'vioscreen_visit', 'vioscreen_vita_iu', 'vioscreen_vita_rae', 'vioscreen_vita_re', 'vioscreen_vitb12', 'vioscreen_vitb6', 'vioscreen_vitc', 'vioscreen_vitd', 'vioscreen_vitd2', 'vioscreen_vitd3', 'vioscreen_vitd_iu', 'vioscreen_vite_iu', 'vioscreen_vitk', 'vioscreen_water', 'vioscreen_weight', 'vioscreen_wgrain', 'vioscreen_whole_grain_servings', 'vioscreen_xylitol', 'vioscreen_zinc', 'vitamin_b_supplement_frequency', 'vitamin_d_supplement_frequency', 'vivid_dreams', 'weekday_sleep_time', 'weekday_wake_time', 'weekend_sleep_time', 'weekend_wake_time', 'weight_cat', 'weight_change', 'weight_kg', 'weight_units', 'whole_eggs', 'whole_grain_frequency'\n",
       "    var: 'taxonomy_0', 'taxonomy_1', 'taxonomy_2', 'taxonomy_3', 'taxonomy_4', 'taxonomy_5', 'taxonomy_6'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save adata to file:\n",
    "# adata.write_h5ad(\"full_table.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata from file:\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e4dc294415b14ea4c0d4d07894c031d87e01c25f832e475b95612eada6e667d4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
